### DISCUSSION TITLE
Taiko playfield scale shouldn't be fixed to 16:9 since many players use a different aspect ratio on purpose

### DISCUSSION THREAD

**Discussion by mouzedrift** [ppy/osu#13047] [Ideas] [answered]
**Describe the bug:** 
playfield scaling for taiko doesn't match stable. some people (including me) use different resolutions/aspect ratio to play taiko because it's what they are used to or it feels more "original". i would suggest the original scale behavior like stable for legacy skins or make it a setting so people don't have to re-learn the game as this makes a big difference for people who didn't play with 16:9 in the first place.

**Comparison below:**
| osu!stable | osu!lazer |
| ----------- | ---------- |
| ![screenshot258](https://user-images.githubusercontent.com/43358824/119873363-c16ff580-bf24-11eb-8b6b-1187b6b45622.jpg) | ![osu_2021-05-27_19-48-30](https://user-images.githubusercontent.com/43358824/119873391-c7fe6d00-bf24-11eb-8795-91db5717472d.png) |

**osu!lazer version:**
2021.524.0

edit: updated comparison images and description to be more accurate.

**Comment by peppy** [comment]
This is intentional currently to keep the playfield in a constant scale across all aspect ratios. If it was to change, you would have less visibility of hitobjects at square aspect ratios – is that what you want?

**Comment by mouzedrift** [comment]
yes, at least for legacy skins.

**Comment by ayyEve** [comment]
There are many people who play 4:3 because of the different play-field. I believe it would be unfair to them to force this change. I would keep the field consistent between resolutions but change based on aspect ratio (at least between 4:3 and 16:9).

**Comment by JoogsWasTaken** [comment]
I'll add my two cents as to why this should be carried over from stable. I played on 16:9, 16:10 and 4:3 for a significant amount of time and I can attest that aspect ratios do make a huge difference in taiko. First off,

> If it was to change, you would have less visibility of hitobjects at square aspect ratios – is that what you want?

Yes, this is absolutely why many prefer 4:3. In my personal experience, the way that 4:3 restricts the amount of hit objects present on the screen makes playing maps with no mods easier to process. Not only that, but with the same window height, objects appear larger when using 4:3 as opposed to any other aspect ratio which helps with reading. Again, speaking from my experience, simply learning and training patterns proved to be much easier on 4:3 because there is simply less to take in and to process, and playing with 4:3 gave me the biggest skill boost in taiko.

On the flipside, I'm aware of players who use exclusively either 16:9, 16:10 or 21:9. A reason being that mods which increase difficulty are easier to learn on a wider aspect ratio as compared to 4:3. I'd argue that both HR and DT are signficantly more difficult to get the hang of when playing with 4:3. And although 4:3 may be comfier in some circumstances as explained above, I haven't heard anyone severely struggling with getting better at taiko on different aspect ratios. Furthermore, there are players who will switch between 4:3 and a wider aspect ratio depending on the mods they plan on using.

So tl;dr I guess is that in stable, many taiko players have their preferred aspect ratios for good reasons. The main reason, as people in this discussion have said before me, is that players who are used to 4:3 will absolutely struggle to regain their skill if they're forced to play on a field that is effectively scaled to fit a 16:9 aspect ratio, even when the window around it is not.

Edit (2021/05/31): [Here's a video comparing the playfields between stable (b20210520.2) and lazer (2021.524.0) at 1024x768.](https://streamable.com/eck1uc) This should be the exact behavior described in the original discussion starter. Song is 140 BPM, mapped with 1.0 SV.

**Reply by smoogipoo** [reply]
Would having playfield covers be an acceptable solution? And/or is the speed of hitobjects also a large factor reducing adaptability by 4:3 players?

**Reply by JoogsWasTaken** [reply]
Just fyi, I updated my original comment with a link to a video highlighting the difference between stable and lazer as of writing. Think this should clear things up for people who join this thread later.

> Would having playfield covers be an acceptable solution?

I spent the evening trying to recreate this behavior to my understanding. What I'm guessing is that you're suggesting to cover the playfield in lazer to match the behavior in stable if an aspect ratio other than 16:9 is selected? Correct me if I'm wrong.

I chose 1024x768 as an example just like in the video that I linked in my original post and if I got that correctly, then what you're suggesting would look [something like this](https://streamable.com/uq4c5g). If I were to resize the playfield in lazer to fill the entire window width, then [stable and lazer would indeed match up](https://streamable.com/b75ddw).

However, covering the playfield would come with its own issues which is, in my understanding, what your second question is about.

> And/or is the speed of hitobjects also a large factor reducing adaptability by 4:3 players?

Yes, covering the playfield would make the hit objects smaller and travel slower across the screen. So if this were to be implemented in lazer, players who previously played on 4:3 in stable would still have to get used to a different playfield. And although I'd argue that the noticeable difference while playing would be smaller compared to having to learn how to play on a forced 16:9 aspect ratio, it's still not a change that I feel would be accepted by most players. This is entirely subjective of course. I obviously haven't had the time yet to figure out just how much smaller and slower hit objects appear between stable and lazer at the same resolution with a 4:3 aspect ratio.

But I think covering the playfield introduces another problem which is mostly centered around how the hidden modification is currently handled in stable. When applying HD while playing on an aspect ratio other than 4:3, [the playfield is covered up so it's forced to 4:3](https://i.imgur.com/neLIZ4o.png). This is the exact opposite to what you're suggesting, except it is actually implemented in stable. I saw that HD is currently not implemented for lazer in taiko, so depending on how HD will be implemented later in lazer's lifecycle, this could be another pitfall.

So again tl;dr no, I don't think covering the playfield would be an acceptable solution here, because the difference in speed and size of hit objects at the same resolution (given a 4:3 aspect ratio) between stable and lazer would turn players off. Also, care should be taken when HD is eventually implemented in lazer for taiko, because stable currently implements it in a way that acts in the exact opposite way as to what you're suggesting.

**Comment by chayleaf** [comment]
In a way, this is the same as people that can only play high AR in std. As a 16:9 player, I do consider 4:3 cheat-ish - it's somewhat like HR without the OD change, so it makes everything easier to acc and makes dense maps easier to read. The only problem is it's impossible to counter completely, you can achieve the same effect without changing anything in-game by simply closing part of the screen with a towel or something else. That's why I think it's fine to allow such resolutions, even if it is unfair - simply because it can be easily achieved even without such a feature. Of course you could instead say covering parts of the screen is now considered cheating and not allowed, but that would be impossible to verify, besides liveplays.

**Comment by chayleaf** [comment]
- 4:3 resolution DOES change game balance, because it decreases note density. It's not "just" a preference, people who use that resolution simply can't play 16:9. It's like permanently increasing AR in std.

- That's why I think this should be implemented as a system mod like touchscreen to make sure replays accurately represent the player experience and make sure the data is available to the pp system (strictly speaking, you can't make sure the player sees the entire playfield, but you can at least make sure the player doesn't see more than the playfield)

- The actual place to input the playfield scale should probably be the game settings.

- The formula to calculate the playfield scale to simulate x:y resolution in stable is 1/(x:y-1/3)

- HR mod in stable accounts for the resolution in the wrong way (by comparing screen dimensions instead of playfield dimensions), so it shouldn't be perfectly simulated.
  - Currently, HR mod in lazer is hardcoded to 16:9. 4:3 makes notes bigger and that makes reading easier, but since playfield scale changes balance even in NM, I don't see a reason to disallow changing balance here. Basically, SV scaling needs to be implemented - and players should decide what is or isn't fair themselves.

- HD mod is 4:3 by default - it's a really annoying special case, more so than HR
   - In stable, HD accounts for the base SV instead of post-mod (EZ/HR) SV, should that be reproduced?
   - It is especially annoying when considering HDHR. How should it look? Should it look exactly like the HR-only playfield but with HD applied?
   - How should the HD playfield configuration look like? Just like in HR, I think it's ok to let players configure the playfield if the configuration is visible in the replays - but the thing is, HD is 4:3 by default, it's balanced for 4:3, and changing the playfield scale heavily changes the mod balance. That's why it needs to be a special case of resolution configuration. Should there simply be a separate resolution setting for HD? Should there be a toggle "Always fix HD resolution to 4:3"?

**Reply by Firmatorenio** [reply]
Making 4:3 a ranked mod with no score multiplier would significantly simplify coding HR and HD compared to coding them for different resolutions, as well as directly represent the players' configuration in the replay data, I personally think that could be a valid solution.

**Reply by JoogsWasTaken** [reply]
> 4:3 resolution DOES change game balance, because it decreases note density. [...] That's why I think this should be implemented as a system mod [...] to make sure replays accurately represent the player experience [...]

What about 5:4 then? It's an option in stable, e.g. 1280x1024, and it decreases note density even more. We're not just talking about 16:9 and 4:3 here. Taiko players use a wide array of aspect ratios. It just so happens that 4:3 and 16:9 are the most common ones. [There's a list of commonly used aspect ratios on Wikipedia](https://en.wikipedia.org/wiki/Display_aspect_ratio). I personally know people who play on 16:10, 21:9 and 3:2, and I bet you're gonna find a few more if you ask more players.

If you're gonna go for a "system mod" approach, you'll have to find a "magic" aspect ratio at which people agree that the note density is low enough that it warrants to be declared a significant difference in playstyle. This won't work since custom resolutions (by editing config files) are possible in stable and (currently) in lazer, and players can just work their way around this barrier easily.

> HR mod in stable accounts for the resolution in the wrong way (by comparing screen dimensions instead of playfield dimensions), so it shouldn't be perfectly simulated.

I don't quite get what you mean by that. Maybe it's also because of my inexperience with HR. Can you illustrate this somehow?

> HD mod is 4:3 by default - it's a really annoying special case, more so than HR

I agree with you on that. The way HD works in stable feels like a dirty hack, and probably has been since it's been implemented. I don't think that covering the playfield should be something that should be carried over to lazer once HD is implemented for taiko. In my opinion, people who want to keep the feel of HD consistent with their experience coming from stable should just pick a 4:3 aspect ratio in lazer if they intend on playing with HD.

Also, I absolutely do believe that HD using the base SV should stay. Correct me if I'm wrong, but in standard, with HD there's a consistent time window before an object is meant to be hit. If that were the case for taiko, then the horizontal position at which objects would fade out would vary depending on the active SV at the hit object's offset. This would be painful to read, as well as too big of a break from how HD is implemented in stable currently.

**Comment by mangomizer** [comment]
As an immediate stopgap to allow accurate emulation for different resolutions on lazer, I would advise allowing the scroll speed adjust (from "difficulty adjust" mod) to be incrementable in steps of 0.01 (currently 0.05).

This will require a more elegant fix in the future, but in the short term this will open the doors for all non-16:9 users to lazer (who for the most part would have found the game to be unplayable otherwise), and thus allowing more manpower in identifying issues and finding solutions for the gamemode.

In support of this, I will type up and spread a rough guide for temporary res fixes once this is complete (to distribute).

**Comment by smoogipoo** [comment]
As of https://github.com/ppy/osu/pull/14105 and https://github.com/ppy/osu/pull/14106, the classic mod matches osu!stable's scroll speed and the 16:9 playfield size.

I'm undecided in making the playfield size match at 4:3 for the time being.

**Comment by zachmanthethird** [comment]
I wish I could provide an answer, but I don't really have a solution. However, after today's community meeting, this HAS to be addressed.
This is the biggest reason why most osu!taiko players have not transitioned. Scores being worth anything in any form is going to instantly turn away the majority of the active taiko community.
Additional issues include #14798 regarding HD implementation and #14794 regarding the "Classic" mod. Flashlight is also an issue, but without the two prior issues resolved, there's no point in discussing it yet.

**Reply by peppy** [reply]
Are you confused or am I confused? How is "scores being worth anything" going to "turn away community" due to this issue? As far as I can tell, users which want to play with the previous behaviour can still get it on stable, and the current lazer behaviour is *more* balanced if anything (as in, a user cannot adjust visible area by changing their aspect ratio).

Could you clarify what you are implying?

**Reply by zachmanthethird** [reply]
> As far as I can tell, users which want to play with the previous behaviour can still get it on stable

That's exactly the problem. **If** stable and lazer leaderboards are to be combined in any way, this will enforce the fact that stable and lazer gameplay are the same (see https://github.com/ppy/osu/issues/14794#issuecomment-1059922788). They are not. Therefore, they should not be treated the same way and they should not be worth the same amount of pp.
Let's take HD for example. In stable, HD is locked to 4:3, while in lazer, HD is locked to 16:9. Let's ignore the fact that HD+Classic doesn't match stable, an issue I'm not planning to open until #14794 is resolved. **HD in lazer is objectively more difficult than HD in stable.** While I personally do not agree, some in the community believe that HD in lazer should be worth the same as HDHR in stable. (Edit: I do have proof of this via Discord DMs, but do not think it's relevant here.) With its current implementation, comparing HD in these two clients is illogical. Doing so will actively incentivize HD players away from lazer.
By allowing lazer scores to be worth ranked score and pp, it's reflexively telling the community that osu!taiko gameplay is finished and equal to stable. This is simply not true due to many of the discussions I've opened recently in response to the previous community meeting.

**Reply by peppy** [reply]
Only classic mod is supposed to match. When classic mod is not applied, the gameplay mechanics are intentionally different. I think you may be missing that detail?

**Reply by Chupalika** [reply]
First of all, classic mod is actually not matching stable behavior (#14794 as zach already linked)

Second of all, I think what we're trying to say is that *because* the gameplay mechanics are intentionally different (without classic mod), the pp algorithm should be different, or at least take that into account somehow.

**Reply by chayleaf** [reply]
As of now, Lazer gives no pp, so any inconsistencies are okay. When Lazer will be ready for ranked plays, the community can come up with a pp system for it. Classic mod will match - but it doesn't match yet, because Lazer isn't feature complete compared to stable yet.

**Reply by Lougui** [reply]
In my opinion this is one of the worst things in lazer... If you can not play 4:3 while getting pp in a ranked map, you are going to lose about 1/3 of all taiko players due to the fact that leaving the game you love because you can't play it anymore is actually a thing... I have seen many people being mad about the lazer updates to taiko before, like 3/4 of the people I asked (~200) in the range between 1 and 1.5k actually said they would play less, and if nothing changed actually quit the game... (just on the 4:3 subject)
In addition to that, you can't make a pp system for lazer... just look at how long it takes for us to have a reworked system... you can't just say "oh well then... let's make a new one for lazer!". Just leave 4:3 players alone, let them play the game and... come on... this is not even fair for them, like you can't balance the game by  forcing 1/3 to leave... that's not how it works...

**Comment by frenzibyte** [comment]
I believe it's safe to say the remaining part of this thread is tracked in https://github.com/ppy/osu/issues/14794, since they both focus on the visibility of hitobjects on different aspect ratios (as demonestrated in https://github.com/ppy/osu/issues/14794#issuecomment-1061069674). And osu!(lazer)'s current behaviour is intentional so this won't change outside the scope of "Classic" mod (unless we're deciding this to discuss the current hitobjects visibility range in osu!(lazer)).

**Comment by peppy** [comment, marked as answer]
This has been implemented via https://github.com/ppy/osu/pull/21883 for now.

Also documented at https://github.com/ppy/osu/wiki/Gameplay-differences-from-osu!stable#osutaiko-gamefield-aspect-ratio-is-limited.
